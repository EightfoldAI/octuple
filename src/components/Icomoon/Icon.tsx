import React, { FC, useContext } from 'react';
import { mergeClasses } from '../../shared/utilities';
import { IcomoonProps, IcomoonSize } from './';
import IcomoonReact from 'icomoon-react';

import styles from './icon.module.scss';

const IconSetContext = React.createContext<object | null>(null);

export const IcomoonProvider = IconSetContext.Provider;

export const Icomoon: FC<IcomoonProps> = ({
    ariaHidden = false,
    classNames,
    color,
    iconName,
    id,
    role = 'presentation',
    size = IcomoonSize.Medium,
    'data-test-id': dataTestId,
}) => {
    const iconClassNames: string = mergeClasses([
        classNames,
        styles.iconWrapper,
    ]);

    return (
        <span
            data-test-id={dataTestId}
            aria-hidden={ariaHidden}
            className={iconClassNames}
            id={id}
            role={role}
        >
            <IconSetContext.Consumer>
                {(iconSet) => (
                    <IcomoonReact
                        iconSet={iconSet}
                        icon={iconName}
                        size={size}
                        color={color}
                    />
                )}
            </IconSetContext.Consumer>
        </span>
    );
};
