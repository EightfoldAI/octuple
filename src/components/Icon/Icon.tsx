'use client';

import React, { FC } from 'react';
import IcomoonReact from 'icomoon-react';

import { mergeClasses } from '../../shared/utilities';
import { Icon as MdiIcon } from '@mdi/react';
import { IconProps, IconSize } from './index';

import styles from './icon.module.scss';
import { useConfig } from '../ConfigProvider';

export const Icon: FC<IconProps> = ({
  ariaHidden = false,
  classNames,
  color,
  description,
  horizontal,
  id,
  path,
  role = 'presentation',
  rotate,
  size = IconSize.Medium,
  spin,
  style,
  title,
  vertical,
  'data-test-id': dataTestId,
  icomoonIconName,
}) => {
  const { icomoonIconSet } = useConfig();

  const iconClassNames: string = mergeClasses([classNames, styles.iconWrapper]);

  if (!icomoonIconName && !path) {
    return null;
  }

  const iconComponent = icomoonIconName ? (
    <IcomoonReact
      iconSet={icomoonIconSet}
      size={size}
      color={color}
      icon={icomoonIconName}
      role="presentation"
    />
  ) : (
    <MdiIcon
      color={color}
      description={description}
      horizontal={horizontal}
      path={path}
      rotate={rotate}
      size={size}
      title={title}
      vertical={vertical}
      spin={spin}
    />
  );

  return (
    <span
      data-test-id={dataTestId}
      aria-hidden={ariaHidden}
      className={iconClassNames}
      id={id}
      role={role}
      style={style ? style : null}
    >
      {iconComponent}
    </span>
  );
};

Icon.displayName = 'Icon';
