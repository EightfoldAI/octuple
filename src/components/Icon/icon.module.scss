.iconWrapper {
    display: inline-block;
    margin: 0;
    padding: 0;

    svg {
        vertical-align: middle;
    }
}