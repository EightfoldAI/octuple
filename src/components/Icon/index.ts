export * from './Icon';
export * from './Icon.types';
export * from './mdi';
