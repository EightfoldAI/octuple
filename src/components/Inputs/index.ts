export * from './Input.types';
export * from './TextArea/TextArea';
export * from './TextInput/TextInput';
export * from './SearchBox/SearchBox';
