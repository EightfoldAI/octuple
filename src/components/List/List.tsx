'use client';

import React, { Key, useEffect, useRef, useState } from 'react';
import { ListProps } from './List.types';
import { ExternalListItem } from './ListItem';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import {
  eventKeys,
  focusable,
  mergeClasses,
  SELECTORS,
} from '../../shared/utilities';

import styles from './list.module.scss';

export const List = <T extends any>({
  additionalItem,
  disableArrowKeys = false,
  disableKeys = false,
  items,
  footer,
  layout = 'vertical',
  renderAdditionalItem,
  renderItem,
  rowKey,
  header,
  classNames,
  style,
  itemClassNames,
  itemStyle,
  listClassNames,
  listStyle,
  listType = 'ul',
  role,
  itemProps,
  getItem,
  ...rest
}: ListProps<T>) => {
  const htmlDir: string = useCanvasDirection();

  const containerClasses: string = mergeClasses([
    styles.listContainer,
    listClassNames,
    { [styles.vertical]: layout === 'vertical' },
  ]);
  const [focusIndex, setFocusIndex] = useState<number | null>(null);
  const itemRefs: React.MutableRefObject<HTMLElement[]> = useRef<
    (HTMLElement | null)[]
  >([]);

  // Update the focused item when the items or focusIndex change.
  useEffect(() => {
    // Ensure the focused item index is within bounds.
    if (focusIndex !== null && items && focusIndex >= items.length) {
      setFocusIndex(items.length - 1);
    }
  }, [focusIndex, items]);

  const handleItemKeyDown = (
    event: KeyboardEvent | React.KeyboardEvent<HTMLElement>,
    index: number,
    external: boolean = false
  ): void => {
    if (disableArrowKeys || disableKeys) {
      return;
    }
    const arrowDown: boolean = event?.key === eventKeys.ARROWDOWN;
    const arrowLeft: boolean = event?.key === eventKeys.ARROWLEFT;
    const arrowRight: boolean = event?.key === eventKeys.ARROWRIGHT;
    const arrowUp: boolean = event?.key === eventKeys.ARROWUP;
    const arrowDecrement: boolean = htmlDir === 'rtl' ? arrowRight : arrowLeft;
    const arrowIncrement: boolean = htmlDir === 'rtl' ? arrowLeft : arrowRight;
    const end: boolean = event?.key === eventKeys.END;
    const home: boolean = event?.key === eventKeys.HOME;
    if (
      ((arrowDown || arrowUp) && layout === 'vertical') ||
      ((arrowDecrement || arrowIncrement) && layout === 'horizontal')
    ) {
      event?.preventDefault();
      setFocusIndex(index);
      if (itemRefs.current[index]) {
        const getFocusableElements = (): HTMLElement[] => {
          const step: number =
            (arrowDown && layout === 'vertical') ||
            (arrowIncrement && layout === 'horizontal')
              ? 1
              : -1;
          let nextIndex: number = index + step;
          const additionalItemIndex: number = items ? items.length : 0;
          if (
            (renderAdditionalItem &&
              ((arrowDown && layout === 'vertical') ||
                (arrowIncrement && layout === 'horizontal')) &&
              nextIndex === additionalItemIndex) ||
            (((arrowUp && layout === 'vertical') ||
              (arrowDecrement && layout === 'horizontal')) &&
              nextIndex === additionalItemIndex + 1)
          ) {
            return [
              ...(itemRefs.current[additionalItemIndex]?.querySelectorAll(
                SELECTORS
              ) as unknown as HTMLElement[]),
            ].filter((el: HTMLElement) => focusable(el));
          }
          while (
            (nextIndex >= 0 && nextIndex <= additionalItemIndex) ||
            (nextIndex > additionalItemIndex && nextIndex < items.length)
          ) {
            const nextItem: HTMLElement | null = itemRefs.current[nextIndex];
            if (nextItem && focusable(nextItem)) {
              const nextFocusableItem: HTMLElement | null = external
                ? nextItem.parentElement
                : nextItem;
              return [
                ...(nextFocusableItem.querySelectorAll(
                  SELECTORS
                ) as unknown as HTMLElement[]),
              ].filter((el: HTMLElement) => focusable(el));
            }
            nextIndex += step;
          }
          return [];
        };
        const focusableElements: HTMLElement[] = getFocusableElements();
        focusableElements?.[0]?.focus();
      }
    }
    if (end || home) {
      const focusElement = (index: number) => {
        event?.preventDefault();
        setFocusIndex(index);
        if (itemRefs.current[index]) {
          const item: HTMLElement | null = external
            ? itemRefs.current[index].parentElement
            : itemRefs.current[index];
          const getFocusableElements = (): HTMLElement[] => {
            return [
              ...(item.querySelectorAll(SELECTORS) as unknown as HTMLElement[]),
            ].filter((el: HTMLElement) => focusable(el));
          };
          const focusableElements: HTMLElement[] = getFocusableElements();
          const focusableElement: HTMLElement | null = focusableElements?.[0];
          focusableElement?.focus();
        }
      };

      if (end) {
        if (renderAdditionalItem) {
          focusElement(items.length);
        } else {
          focusElement(items.length - 1);
        }
      }

      if (home) {
        focusElement(0);
      }
    }
  };

  const itemClasses: string = mergeClasses([styles.listItem, itemClassNames]);

  const getHeader = (): JSX.Element => <>{header}</>;

  const getFooter = (): JSX.Element => <>{footer}</>;

  const getAdditionalItem = (): JSX.Element => {
    const additionalItemIndex: number = items ? items.length : 0;
    const itemRef = (el: HTMLElement | null) => {
      itemRefs.current[additionalItemIndex] = el;
    };
    return (
      <li
        {...itemProps}
        key={getItemKey(additionalItem, additionalItemIndex)}
        className={itemClasses}
        onKeyDown={(event: React.KeyboardEvent<HTMLElement>) =>
          handleItemKeyDown(event, additionalItemIndex)
        }
        ref={itemRef}
        style={itemStyle}
      >
        {renderAdditionalItem?.(additionalItem)}
      </li>
    );
  };

  const getItemKey = (item: T, index: number): Key => {
    if (typeof rowKey === 'function') {
      return rowKey(item) as Key;
    } else if (rowKey) {
      return item[rowKey];
    }
    return `oc-list-item-${index}`;
  };

  const _getItem = (
    item: T,
    index: number,
    itemRef: (el: HTMLElement | null) => void
  ): JSX.Element => (
    <li
      {...itemProps}
      key={getItemKey(item, index)}
      className={itemClasses}
      onKeyDown={(event: React.KeyboardEvent<HTMLElement>) =>
        handleItemKeyDown(event, index)
      }
      ref={itemRef}
      style={itemStyle}
    >
      {renderItem(item)}
    </li>
  );

  const externalItem = (
    node: React.ReactNode,
    index: number,
    itemRef: (el: HTMLElement | null) => void
  ): JSX.Element => {
    if (!node) {
      return null;
    }
    const item: React.ReactElement<
      HTMLElement,
      string | React.JSXElementConstructor<any>
    > = React.Children.only(node) as React.ReactElement<HTMLElement>;
    return (
      <ExternalListItem
        handleItemKeyDown={handleItemKeyDown}
        id={item?.props?.id}
        index={index}
        item={item}
        itemRefs={itemRefs}
        itemRef={itemRef}
      />
    );
  };

  const getItems = (): React.ReactNode[] =>
    items.map((item: T, index) => {
      const itemRef = (el: HTMLElement | null) => {
        itemRefs.current[index] = el;
      };
      return (
        externalItem?.(getItem?.(item, index), index, itemRef) ??
        _getItem(item, index, itemRef)
      );
    });

  return (
    <div {...rest} className={classNames} style={style}>
      {getHeader()}
      {listType === 'ul' && (
        <ul role={role} className={containerClasses} style={{ ...listStyle }}>
          {getItems()}
          {!!renderAdditionalItem && getAdditionalItem()}
        </ul>
      )}
      {listType === 'ol' && (
        <ol role={role} className={containerClasses} style={{ ...listStyle }}>
          {getItems()}
          {!!renderAdditionalItem && getAdditionalItem()}
        </ol>
      )}
      {getFooter()}
    </div>
  );
};
