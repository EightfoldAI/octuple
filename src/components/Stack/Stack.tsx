'use client';

import React, { FC, Ref } from 'react';
import { mergeClasses } from '../../shared/utilities';
import { StackBreakpoint, StackProps } from './Stack.types';
import { useMatchMedia } from '../../octuple';
import { Breakpoints } from '../../hooks/useMatchMedia';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';

import styles from './stack.module.scss';

export const Stack: FC<StackProps> = React.forwardRef(
  (
    {
      fullWidth: defaultFullWidth,
      direction: defaultDirection = 'horizontal',
      justify: defaultJustify,
      inline: defaultInline,
      align: defaultAlign,
      wrap: defaultWrap,
      gap: defaultGap,
      flexGap: defaultFlexGap,
      style = {},
      classNames,
      breakpoints = {},
      children,
      ...rest
    },
    ref: Ref<HTMLDivElement>
  ) => {
    const xSmallScreenActive: boolean = useMatchMedia(Breakpoints.XSmall);
    const smallScreenActive: boolean = useMatchMedia(Breakpoints.Small);
    const mediumScreenActive: boolean = useMatchMedia(Breakpoints.Medium);
    const largeScreenActive: boolean = useMatchMedia(Breakpoints.Large);

    const htmlDir: string = useCanvasDirection();

    const breakPointConfigurationList: Array<[StackBreakpoint, boolean]> = [
      ['large', largeScreenActive],
      ['medium', mediumScreenActive],
      ['small', smallScreenActive],
      ['xsmall', xSmallScreenActive],
    ];

    const activeBreakPointStackPropsIndex =
      breakPointConfigurationList.findIndex((breakPointConfiguration) => {
        return breakPointConfiguration[1];
      });

    const activeBreakPointStackProps =
      activeBreakPointStackPropsIndex > -1
        ? breakpoints[
            breakPointConfigurationList[activeBreakPointStackPropsIndex][0]
          ]
        : {};

    const resolvedStackIntrinsicProps = {
      ...{
        fullWidth: defaultFullWidth,
        direction: defaultDirection,
        justify: defaultJustify,
        inline: defaultInline,
        align: defaultAlign,
        wrap: defaultWrap,
        gap: defaultGap,
        flexGap: defaultFlexGap,
      },
      ...(activeBreakPointStackProps || {}),
    };

    const { fullWidth, direction, justify, inline, align, wrap, gap, flexGap } =
      resolvedStackIntrinsicProps;

    const stackClassName: string = mergeClasses([
      styles.stack,
      classNames,
      { [styles.inline]: inline },
      { [styles.fullWidth]: fullWidth },
      { [styles.vertical]: direction === 'vertical' },
      { [styles.horizontal]: direction === 'horizontal' },
      { [styles.stackRtl]: htmlDir === 'rtl' },
      styles[gap],
      { [styles[`gap-${flexGap}`]]: flexGap },
    ]);

    return (
      <div
        ref={ref}
        className={stackClassName}
        style={{
          alignItems: align,
          justifyContent: justify,
          flexWrap: wrap,
          ...style,
        }}
        {...rest}
      >
        {children}
      </div>
    );
  }
);
