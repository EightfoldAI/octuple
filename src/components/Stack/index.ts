export * from './Stack.types';
export * from './Stack';
