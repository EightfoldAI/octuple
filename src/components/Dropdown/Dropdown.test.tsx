import React, { useEffect, useRef, useState } from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import MatchMediaMock from 'jest-matchmedia-mock';
import { Dropdown } from './';
import {
  Button,
  ButtonIconAlign,
  ButtonTextAlign,
  ButtonWidth,
} from '../Button';
import { Icon, IconName } from '../Icon';
import { List } from '../List';
import { Stack } from '../Stack';
import { TextInput } from '../Inputs';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';
import '@testing-library/jest-dom';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia: any;

interface User {
  name: string;
  icon: IconName;
}

const sampleList: User[] = [1, 2, 3].map((i) => ({
  name: `User profile ${i}`,
  icon: IconName.mdiAccount,
}));

const Overlay = () => (
  <List<User>
    items={sampleList}
    renderItem={(item) => (
      <Button
        text={item.name}
        alignText={ButtonTextAlign.Left}
        buttonWidth={ButtonWidth.fill}
        data-testid={item.name}
        iconProps={{
          path: item.icon,
        }}
        style={{
          margin: '4px 0',
        }}
      />
    )}
  />
);

const dropdownProps: object = {
  trigger: 'click',
  classNames: 'my-dropdown-class',
  style: {},
  dropdownClassNames: 'my-dropdown-class',
  dropdownStyle: {
    color: 'red',
  },
  placement: 'bottom-start',
  overlay: Overlay(),
  offset: 0,
  positionStrategy: 'absolute',
  disabled: false,
  closeOnDropdownClick: true,
  portal: false,
};

const DropdownComponent = (): JSX.Element => {
  const [visible, setVisibility] = useState(false);

  return (
    <Dropdown
      {...dropdownProps}
      onVisibleChange={(isVisible) => setVisibility(isVisible)}
    >
      <Button
        alignIcon={ButtonIconAlign.Right}
        text={'Dropdown menu test'}
        iconProps={{
          path: IconName.mdiChevronDown,
          rotate: visible ? 180 : 0,
        }}
        id="test-button-id"
        data-testid="dropdown-reference"
      />
    </Dropdown>
  );
};

const ComplexDropdownComponent = (): JSX.Element => {
  const inputRef: React.MutableRefObject<HTMLInputElement> =
    useRef<HTMLInputElement>(null);
  const [visible, setVisibility] = useState(false);

  return (
    <Dropdown
      {...dropdownProps}
      ariaRef={inputRef}
      onVisibleChange={(isVisible) => setVisibility(isVisible)}
    >
      <div id="wrapper">
        <Stack direction="horizontal" flexGap="xl">
          <Icon path={IconName.mdiAccount} />
          <TextInput
            ref={inputRef}
            placeholder={'Select'}
            iconProps={{
              path: IconName.mdiChevronDown,
              rotate: visible ? 180 : 0,
            }}
            role="combobox"
            data-testid="test-input-id"
          />
        </Stack>
      </div>
    </Dropdown>
  );
};

const ExternalElementDropdownComponent = (): JSX.Element => {
  const [visible, setVisibility] = useState(false);

  return (
    <Stack direction="horizontal" flexGap="xxl">
      <Button
        alignIcon={ButtonIconAlign.Right}
        checked={visible}
        data-testid="test-external-button-id"
        onClick={() => setVisibility(!visible)}
        text={'External Control'}
        toggle
      />
      <Dropdown
        {...dropdownProps}
        visible={visible}
        onVisibleChange={(isVisible) => setVisibility(isVisible)}
      >
        <Button
          alignIcon={ButtonIconAlign.Right}
          text={'Dropdown menu test'}
          iconProps={{
            path: IconName.mdiChevronDown,
            rotate: visible ? 180 : 0,
          }}
          data-testid="test-button-id"
        />
      </Dropdown>
    </Stack>
  );
};

const filterOverlay1 = () => (
  <List<User>
    items={sampleList}
    renderItem={(item) => (
      <Button
        text={'Filter 1' + ' ' + item.name}
        alignText={ButtonTextAlign.Left}
        buttonWidth={ButtonWidth.fill}
        iconProps={{
          path: item.icon,
        }}
        style={{
          margin: '4px 0',
        }}
      />
    )}
  />
);

const filterOverlay2 = () => (
  <List<User>
    items={sampleList}
    renderItem={(item) => (
      <Button
        text={'Filter 2' + ' ' + item.name}
        alignText={ButtonTextAlign.Left}
        buttonWidth={ButtonWidth.fill}
        iconProps={{
          path: item.icon,
        }}
        style={{
          margin: '4px 0',
        }}
      />
    )}
  />
);

const filterOverlay3 = () => (
  <List<User>
    items={sampleList}
    renderItem={(item) => (
      <Button
        text={'Filter 3' + ' ' + item.name}
        alignText={ButtonTextAlign.Left}
        buttonWidth={ButtonWidth.fill}
        iconProps={{
          path: item.icon,
        }}
        style={{
          margin: '4px 0',
        }}
      />
    )}
  />
);

const AdvancedVisibilityToggleDropdownComponent = (): JSX.Element => {
  const [visibleQuickFilter, setVisibleQuickFilter] = useState<{
    [x: string]: boolean;
  }>({});
  const [filterOneVisible, setFilterOneVisibility] = useState(false);
  const [filterTwoVisible, setFilterTwoVisibility] = useState(false);
  const [filterThreeVisible, setFilterThreeVisibility] = useState(false);
  const [filterThreeText, setFilterThreeText] = useState('Filter three closed');

  const toggleQuickFilterVisibility = ({
    key,
    value,
  }: {
    key: string;
    value: boolean;
  }) => {
    setVisibleQuickFilter({ [key]: value });
  };

  useEffect(() => {
    if ((visibleQuickFilter as any)?.['filter one']) {
      setFilterOneVisibility(true);
      setFilterTwoVisibility(false);
      setFilterThreeVisibility(false);
    } else if ((visibleQuickFilter as any)?.['filter two']) {
      setFilterOneVisibility(false);
      setFilterTwoVisibility(true);
      setFilterThreeVisibility(false);
    } else if ((visibleQuickFilter as any)?.['filter three']) {
      setFilterOneVisibility(false);
      setFilterTwoVisibility(false);
      setFilterThreeVisibility(true);
    }

    setFilterThreeText(
      (visibleQuickFilter as any)?.['filter three']
        ? 'Filter three open'
        : 'Filter three closed'
    );
  }, [
    filterOneVisible,
    filterTwoVisible,
    filterThreeVisible,
    visibleQuickFilter,
  ]);

  return (
    <Stack direction="horizontal" flexGap="l">
      <Dropdown
        {...dropdownProps}
        closeOnDropdownClick={false}
        closeOnOutsideClick={false}
        dropdownClassNames={'my-dropdown-class-1'}
        onVisibleChange={(visible: boolean) => {
          toggleQuickFilterVisibility({
            key: 'filter one',
            value: visible,
          });
        }}
        overlay={filterOverlay1()}
        visible={filterOneVisible}
      >
        <Button
          alignIcon={ButtonIconAlign.Right}
          data-testid="test-button-one-id"
          text={'Filter one'}
          iconProps={{
            path: IconName.mdiChevronDown,
          }}
        />
      </Dropdown>
      <Dropdown
        {...dropdownProps}
        closeOnDropdownClick={false}
        closeOnOutsideClick={false}
        dropdownClassNames={'my-dropdown-class-2'}
        onVisibleChange={(visible: boolean) => {
          toggleQuickFilterVisibility({
            key: 'filter two',
            value: visible,
          });
        }}
        overlay={filterOverlay2()}
        visible={filterTwoVisible}
      >
        <Button
          alignIcon={ButtonIconAlign.Right}
          data-testid="test-button-two-id"
          text={'Filter two'}
          iconProps={{
            path: IconName.mdiChevronDown,
          }}
        />
      </Dropdown>
      <Dropdown
        {...dropdownProps}
        closeOnDropdownClick={false}
        closeOnOutsideClick={false}
        dropdownClassNames={'my-dropdown-class-3'}
        onVisibleChange={(visible: boolean) => {
          toggleQuickFilterVisibility({
            key: 'filter three',
            value: visible,
          });
        }}
        overlay={filterOverlay3()}
        visible={filterThreeVisible}
      >
        <Button
          alignIcon={ButtonIconAlign.Right}
          data-testid="test-button-three-id"
          text={filterThreeText}
          iconProps={{
            path: IconName.mdiChevronDown,
          }}
        />
      </Dropdown>
    </Stack>
  );
};

describe('Dropdown', () => {
  beforeAll(() => {
    matchMedia = new MatchMediaMock();
  });

  afterEach(() => {
    matchMedia.clear();
  });

  test('Should render a dropdown button', () => {
    const { container } = render(<DropdownComponent />);
    const dropdownButton = screen.getByRole('button');
    expect(dropdownButton).toBeTruthy();
    expect(() => container).not.toThrowError();
    expect(container).toMatchSnapshot();
  });

  test('Should show the dropdown options when the button is clicked', async () => {
    render(<DropdownComponent />);
    const dropdownButton = screen.getByRole('button');
    dropdownButton.click();
    await waitFor(() => screen.getByText('User profile 1'));
    const option1 = screen.getByText('User profile 1');
    const option2 = screen.getByText('User profile 2');
    const option3 = screen.getByText('User profile 3');
    expect(option1).toBeTruthy();
    expect(option2).toBeTruthy();
    expect(option3).toBeTruthy();
  });

  test('Should support props such as dropdownClassNames and dropdownStyle', async () => {
    const { container } = render(<DropdownComponent />);
    const dropdownButton = screen.getByRole('button');
    dropdownButton.click();
    await waitFor(() => screen.getByText('User profile 1'));
    expect(container.querySelector('.dropdown-wrapper')?.classList).toContain(
      'my-dropdown-class'
    );
    expect(
      (container.querySelector('.dropdown-wrapper') as HTMLElement).style.color
    ).toContain('red');
  });

  test('Should support cloned element id from its props', async () => {
    render(<DropdownComponent />);
    const dropdownButton = screen.getByRole('button');
    expect(dropdownButton.id).toBe('test-button-id');
  });

  test('Should support ariaRef prop for complex dropdown references', async () => {
    const { container } = render(<ComplexDropdownComponent />);
    const dropdownAriaRef = screen.getByTestId('test-input-id');
    expect(dropdownAriaRef.getAttribute('aria-controls')).toBeTruthy();
    expect(dropdownAriaRef.getAttribute('aria-expanded')).toBe('false');
    expect(dropdownAriaRef.getAttribute('aria-haspopup')).toBe('true');
    expect(dropdownAriaRef.getAttribute('role')).toBe('combobox');
    dropdownAriaRef.click();
    await waitFor(() => screen.getByText('User profile 1'));
    const option1 = screen.getByText('User profile 1');
    expect(option1).toBeTruthy();
    expect(container.querySelector('.dropdown-wrapper')?.classList).toContain(
      'my-dropdown-class'
    );
    expect(dropdownAriaRef.getAttribute('aria-expanded')).toBe('true');
  });

  test('Should support visible prop toggle via external element', async () => {
    const { container } = render(<ExternalElementDropdownComponent />);
    const externalElement = screen.getByTestId('test-external-button-id');
    const referenceElement = screen.getByTestId('test-button-id');
    externalElement.click();
    await waitFor(() => screen.getByText('User profile 1'));
    const option1 = screen.getByText('User profile 1');
    expect(option1).toBeTruthy();
    expect(container.querySelector('.dropdown-wrapper')?.classList).toContain(
      'my-dropdown-class'
    );
    expect(referenceElement.getAttribute('aria-expanded')).toBe('true');
    externalElement.click();
    await waitFor(() =>
      expect(referenceElement.getAttribute('aria-expanded')).toBe('false')
    );
    expect(container.querySelector('.dropdown-wrapper')).toBeFalsy();
    externalElement.click();
    await waitFor(() => screen.getByText('User profile 1'));
    expect(option1).toBeTruthy();
  });

  test('Should support visible prop toggle via advanced implementation', async () => {
    const { container } = render(<AdvancedVisibilityToggleDropdownComponent />);
    const filterOneReferenceElement = screen.getByTestId('test-button-one-id');
    const filterTwoReferenceElement = screen.getByTestId('test-button-two-id');
    const filterThreeReferenceElement = screen.getByTestId(
      'test-button-three-id'
    );
    filterOneReferenceElement.click();
    await waitFor(() => screen.getByText('Filter 1 User profile 1'));
    const filterOneOption = screen.getByText('Filter 1 User profile 1');
    expect(filterOneOption).toBeTruthy();
    expect(container.querySelector('.dropdown-wrapper')?.classList).toContain(
      'my-dropdown-class-1'
    );
    expect(filterOneReferenceElement.getAttribute('aria-expanded')).toBe(
      'true'
    );
    expect(filterTwoReferenceElement.getAttribute('aria-expanded')).toBe(
      'false'
    );
    expect(filterThreeReferenceElement.getAttribute('aria-expanded')).toBe(
      'false'
    );
    filterTwoReferenceElement.click();
    await waitFor(() => screen.getByText('Filter 2 User profile 1'));
    const filterTwoOption = screen.getByText('Filter 2 User profile 1');
    expect(filterTwoOption).toBeTruthy();
    expect(container.querySelector('.dropdown-wrapper')?.classList).toContain(
      'my-dropdown-class-2'
    );
    expect(filterTwoReferenceElement.getAttribute('aria-expanded')).toBe(
      'true'
    );
    expect(filterOneReferenceElement.getAttribute('aria-expanded')).toBe(
      'false'
    );
    expect(filterThreeReferenceElement.getAttribute('aria-expanded')).toBe(
      'false'
    );
    filterThreeReferenceElement.click();
    await waitFor(() => screen.getByText('Filter 3 User profile 1'));
    const filterThreeOption = screen.getByText('Filter 3 User profile 1');
    expect(filterThreeOption).toBeTruthy();
    expect(container.querySelector('.dropdown-wrapper')?.classList).toContain(
      'my-dropdown-class-3'
    );
    expect(filterThreeReferenceElement.getAttribute('aria-expanded')).toBe(
      'true'
    );
    expect(filterOneReferenceElement.getAttribute('aria-expanded')).toBe(
      'false'
    );
    expect(filterTwoReferenceElement.getAttribute('aria-expanded')).toBe(
      'false'
    );
    filterThreeReferenceElement.click();
    await waitFor(() => screen.getByText('Filter three closed'));
    await waitFor(
      () =>
        filterThreeReferenceElement.getAttribute('aria-expanded') === 'false'
    );
    expect(filterOneReferenceElement.getAttribute('aria-expanded')).toBe(
      'false'
    );
    expect(filterTwoReferenceElement.getAttribute('aria-expanded')).toBe(
      'false'
    );
  });

  test('Handles reference element arrow key events correctly', async () => {
    const mockEventKeys = {
      ARROWDOWN: 'ArrowDown',
      ARROWUP: 'ArrowUp',
    };
    const { container, getByTestId } = render(<DropdownComponent />);
    const referenceElement = getByTestId('dropdown-reference');
    act(() => {
      userEvent.type(referenceElement, mockEventKeys.ARROWDOWN);
    });
    await waitFor(() => screen.getByText('User profile 1'));
    const option1 = screen.getByText('User profile 1');
    expect(option1).toBeTruthy();
    act(() => {
      userEvent.type(referenceElement, mockEventKeys.ARROWUP);
    });
    await waitFor(() =>
      expect(referenceElement.getAttribute('aria-expanded')).toBe('false')
    );
    expect(container.querySelector('.dropdown-wrapper')).toBeFalsy();
  });

  test('Focuses the first focusable element when visible', async () => {
    const { getByTestId } = render(<DropdownComponent />);
    const referenceElement = getByTestId('dropdown-reference');
    act(() => {
      userEvent.click(referenceElement);
    });
    await waitFor(() => screen.getByTestId('User profile 1'));
    await waitFor(() =>
      expect(screen.getByTestId('User profile 1').matches(':focus')).toBe(true)
    );
    expect(screen.getByTestId('User profile 1').matches(':focus')).toBe(true);
  });

  test('Focuses the reference element when not visible', async () => {
    const mockEventKeys = {
      ESCAPE: 'Escape',
    };
    const { container, getByTestId } = render(<DropdownComponent />);
    const referenceElement = getByTestId('dropdown-reference');
    act(() => {
      userEvent.click(referenceElement);
    });
    await waitFor(() => screen.getByText('User profile 1'));
    const option1 = screen.getByText('User profile 1');
    act(() => {
      userEvent.type(option1, mockEventKeys.ESCAPE);
    });
    await waitFor(() =>
      expect(referenceElement.getAttribute('aria-expanded')).toBe('false')
    );
    expect(container.querySelector('.dropdown-wrapper')).toBeFalsy();
    expect(referenceElement).toHaveFocus();
  });
});
