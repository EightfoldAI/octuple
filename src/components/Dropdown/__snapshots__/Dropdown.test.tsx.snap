// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Should render a dropdown button 1`] = `
<div>
  <div
    class="my-dropdown-class main-wrapper"
  >
    <button
      aria-controls="dropdown-"
      aria-disabled="false"
      aria-expanded="false"
      aria-haspopup="true"
      class="button button-default button-medium pill-shape icon-right"
      data-testid="dropdown-reference"
      id="test-button-id"
      role="button"
      tabindex="0"
    >
      <span>
        <span
          aria-hidden="false"
          class="icon icon-wrapper"
          role="presentation"
        >
          <svg
            role="presentation"
            style="width: 20px; height: 20px;"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.41,8.58L12,13.17L16.59,8.58L18,10L12,16L6,10L7.41,8.58Z"
              style="fill: currentColor;"
            />
          </svg>
        </span>
        <span
          class="button-text-medium"
        >
          Dropdown menu test
        </span>
      </span>
    </button>
  </div>
</div>
`;
